src/skeleton/read.rs

```diff
@@ -39,20 +39,27 @@ pub(super) fn manifests<P: AsRef<Path>>(
     base_path: &P,
     metadata: Metadata,
 ) -> Result<Vec<ParsedManifest>, anyhow::Error> {
-    let manifest_paths = metadata
+    let mut packages = metadata
         .workspace_packages()
         .iter()
-        .map(|package| package.manifest_path.clone().into_std_path_buf())
-        .chain(
-            metadata
-                .root_package()
-                .map(|p| p.clone().manifest_path.into_std_path_buf())
-                .or_else(|| Some(base_path.as_ref().join("Cargo.toml"))),
-        )
-        .collect::<BTreeSet<_>>();
+        .copied()
+        .chain(metadata.root_package())
+        .map(|p| (Some(p), p.manifest_path.clone().into_std_path_buf()))
+        .collect::<Vec<_>>();
+
+    if metadata.root_package().is_none() {
+        // At the root, there might be a Cargo.toml manifest with a [workspace] section.
+        // However, if this root manifest doesn't contain [package], it is not considered a package
+        // by cargo metadata. Therefore, we have to add it manually.
+        // Workspaces currently cannot be nested, so this should only happen at the root.
+        packages.push((None, base_path.as_ref().join("Cargo.toml")));
+    }
+
+    packages.sort_by(|a, b| a.1.cmp(&b.1));
+    packages.dedup();
```

Comment:
Can't we use a `BTreeSet` directly here?

Reply:
We just to introduce a small struct and derive the relevant traits, which will give us named fields along the way, making the code a bit clearer as well.

Reply:
With the tuple, probably not, since Package is not orderable. With a custom struct, sure :)

Reply:
```rust
struct PackageMetadata {
    package: Option<Package>,
    manifest_path: PathBuf,
}

impl PartialEq<Self> for PackageMetadata {
    fn eq(&self, other: &Self) -> bool {
        self.manifest_path.eq(&other.manifest_path)
    }
}

impl Eq for PackageMetadata {}

impl PartialOrd for PackageMetadata {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        self.manifest_path.partial_cmp(&other.manifest_path)
    }
}

impl Ord for PackageMetadata {
    fn cmp(&self, other: &Self) -> Ordering {
        self.manifest_path.cmp(&other.manifest_path)
    }
}
```
I'm not sure how to make this shorter. Is it worth it?

Reply:
We don't actually care about _how_ they are ordered, we just need them to be ordered in order to have a deterministic output.

So you can derive all traits without bothering with a manual implementation.
IIRC it's enough to put the path as the first field to have them sorted by path and then by the contents of the package.

Reply:
I can't use derive, since Package isn't Ord. Good point with the order not being really important, although for tests it's nice if it's by path I suppose.

Reply:
I see—you could get around it by grabbing the targets here, instead of doing it below.